src/FA2.hs

```diff
@@ -0,0 +1,85 @@
+module FA2 
+  ( tokenMetadataRegistry
+  , addOperator
+  ) where
+
+import Text.Show (Show (..))
+
+import Lorentz
+import Lorentz.Contracts.Spec.FA2Interface hiding (name, sender) 
+
+import Util.Markdown
+import Util.Named
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+
+type Operators = BigMap (Address, Address) ()
+
+type Ledger = BigMap (Address, TokenId) Natural
+
+type StorageTokenMetadata = ("token_metadata" :! BigMap TokenId TokenMetadata)
+
+-- | Storage of the FA2 contract
+data Storage = Storage
+  { sLedger      :: Ledger
+  , sOperators   :: Operators
+  , sTokenMetadata :: StorageTokenMetadata
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (IsoValue, HasAnnotation)
+
+instance TypeHasDoc Storage where
+  typeDocMdDescription =
+    "Contract's storage holding a big_map with all balances and the operators."
+
+tokenMetadataRegistry :: forall store. Entrypoint TokenMetadataRegistryParam store
+tokenMetadataRegistry = do
+  sender
+  push @Mutez (toMutez 0)
+  swap
+  transferTokens # nil # swap # cons # pair
+
+updateOperator :: Entrypoint UpdateOperatorsParam Storage
+updateOperator = do
+  iter $
+    caseT @UpdateOperator
+      ( #cAdd_operator /-> addOperator
+      , #cRemove_operator /-> removeOperator
+      )
+  nil; pair    
```

Comment:
Well, the idea was to not implement `updateOperators` logic here but instead take it from `morley-ledgers`. Specifically, there is
```
updateOperators
  :: HasStorage Storage
  => OwnerOrOperatorTransfer
  -> Var UpdateOperatorsParam
  -> IndigoProcedure
```

`OwnerOrOperatorTransfer` in our case is `OwnerOrOperatorTransfer ("owner_or_operator_transfer" :! ())`. So it's `HasStorage => Var UpdateOperatorsParam -> IndigoProcedure`. We need to turn it into something like `'[ UpdateOperatorsParam, Storage ]  :-> '[ UpdateOperatorsParam, Storage ]`. In order to do that, we should first make `Var Storage` argument explicit, there should be some way.

`IndigoProcedure` is just a synonym for `IndigoM ()` and if you have a `Var a -> Var b -> IndigoProcedure` and you want to obtain a Lorentz `'[b, a] :-> '[b, a]` you can do so with `compileIndigo @2 funcName` and it should work.

So the only tricky thing is to turn implicit storage argument (`HasStorage`) into an explicit one. It was previously discussed in this thread: https://serokell.slack.com/archives/C017ZDQP2N8/p1599566372004000

Reply:
Am I right than, that I should not write all this functions, instead just set all necessary params and then `compileIndigo` with the function from morley-ledgers?

Reply:
> '[ UpdateOperatorsParam, Storage ] :-> '[ UpdateOperatorsParam, Storage ]
Do we need Params on the top of the stack in the end? I don't see any problems with removing it, cause it is not necessary for the other functions, except this one

Reply:
Also a note. All required functions, that are implemented in `morley-ledgers` `FA2Sample.hs` are not exported from there, so I can't use the lattest version of `morley-ledgers`

Reply:
> Am I right than, that I should not write all this functions, instead just set all necessary params and then `compileIndigo` with the function from morley-ledgers?

Since we are told to implement it in Lorentz, it seems better not to use Indigo code in it, so long term your approach is better. But my idea was to start with reusing code from `morley-ledgers` because I expect it to be faster. Since `addOperator` is already implemented in Lorentz, let's keep it. For other methods you can do as you wish:
1. Implement them in Lorentz ← that's what we eventually need, but writing some logic in Lorentz is hard and time consuming.
2. Convert existing Indigo functions to Lorentz and use them ← it seems easier, but will have to be rewritten at some point. Currently there is no Indigo→Lorentz function that does exactly what we want, which adds extra complexity. I think it's better to create an issue about it and ask Indigo team to do it, will do.

> Do we need Params on the top of the stack in the end? 

Maybe we don't, but that's what `compileIndigo` generates. If you don't need something, you can always `drop` it.

> Also a note. All required functions, that are implemented in morley-ledgers FA2Sample.hs are not exported from there, so I can't use the lattest version of morley-ledgers

Right, I wanted to write about it yesterday, but forgot. First of all, you are welcome to make any changes you need in `morley-ledgers`. Apart from exporting some functions, you probably need to generalize them because your storage is different from the one in `FA2Sample`.

Ideally we should patch `morley-ledgers` and import functions from there, but since we need to hurry up I propose not to bother doing that for now and feel free to copy-paste some code. We will decopypaste it later.

Reply:
> Can you please estimate how much you need to finish that?

I think It won't take long, since token_metadata_registry seems finished (with one mistake, that will be corrected now) and updateOperators also will be finished when addOperator is finished. 

> For other methods you can do as you wish:

I think it will be better if this two stay as pure Lorentz functions, because they are near-complete. All other stuff will be transported as indigo functions firstly, and then decopypasted